pkg/scheduler/framework/plugins/clusteraffinity/types.go

```diff
@@ -6,101 +6,298 @@ Licensed under the MIT license.
+		// The property concerns the available capacity of a resource.
+		q, found = cluster.Status.ResourceUsage.Available[corev1.ResourceName(tn)]
+	default:
+		// The property concerns a capacity type that cannot be recognized.
+		return nil, fmt.Errorf("invalid capacity type %s in resource property name %s", cn, name)
+	}
+
+	if !found {
+		// The property concerns a resource that is not present in the resource usage data.
+		//
+		// It cound be that the resource is not available in the cluster; consequently Fleet
+		// does not consider this as an error.
+		return nil, nil
+	}
+	return &q, nil
 }
 
-// AffinityTerms is a "processed" representation of []ClusterSelectorTerms.
-// The terms are `ORed`.
-type AffinityTerms []affinityTerm
+// retrievePropertyValueFrom retrieves a property value, resource or non-resource,
+// from a member cluster.
+//
+// Note that it will return nil if the property is not available for the cluster;
+// the zero value of resource.Quantity, i.e., resource.Quantity{}, is a valid
+// quantity.
+func retrievePropertyValueFrom(cluster *clusterv1beta1.MemberCluster, name string) (*resource.Quantity, error) {
+	// Check if the expression concerns a resource property.
+	var q *resource.Quantity
+	var err error
+	if strings.Contains(name, resourcePropertyNamePrefix) {
+		name, _ := strings.CutPrefix(name, resourcePropertyNamePrefix)
 
-// Matches returns true if the cluster matches one of the terms.
-func (at AffinityTerms) Matches(cluster *clusterv1beta1.MemberCluster) bool {
-	for _, term := range at {
-		if term.Matches(cluster) {
-			return true
+		// Retrieve the property value from the cluster resource usage data.
+		q, err = retrieveResourcePropertyValueFrom(cluster, name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve resource property value for %s from cluster %s: %w", name, cluster.Name, err)
 		}
+	} else {
+		v, found := cluster.Status.Properties[clusterv1beta1.PropertyName(name)]
+		if !found {
+			// The property is not available for the cluster.
+			//
+			// Note that this is not considered an error.
+			return nil, nil
+		}
+		qv, err := resource.ParseQuantity(v.Value)
+		if err != nil {
+			return nil, fmt.Errorf("value %s of property %s from cluster %s is not a valid quantity: %w", v.Value, name, cluster.Name, err)
+		}
+		q = &qv
 	}
-	return false
+	return q, nil
 }
 
-// preferredAffinityTerm is a "processed" representation of PreferredClusterSelector.
-type preferredAffinityTerm struct {
-	affinityTerm
-	weight int32
-}
+// Matches checks if the cluster matches a cluster requirement.
+//
+// This is an extended method for the ClusterSelectorTerm API.
+func (c *clusterRequirement) Matches(cluster *clusterv1beta1.MemberCluster) (bool, error) {
+	// Match the cluster against the label selector.
+	if c.LabelSelector != nil {
+		ls, err := metav1.LabelSelectorAsSelector(c.LabelSelector)
+		if err != nil {
+			return false, fmt.Errorf("failed to parse label selector: %w", err)
```

Comment:
I wonder why we don't wrap errors like the rest of the controllers do?

Reply:
Oh, that's because it will be wrapped at the top level by the scheduler (e.g., line 486 of framework.go).